src/Orleans/Utils/TimerWheel.cs

```diff
@@ -0,0 +1,128 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Orleans.Runtime;
+
+namespace Orleans.Runtime
+{
+    // Removes the need in timer per item, optimized for singlethreaded consumption. 
+    internal class TimerWheel<T> : IDisposable where T : ITimebound
+    {
+        private const int MaxItemsPerLock = 10;
+        private readonly Queue<T> _entries = new Queue<T>();
+        private readonly InterlockedExchangeLock _queueLock = new InterlockedExchangeLock();
+        private readonly Timer _queueChecker;
+        private readonly Func<T, bool> _shouldDequeue;
+        private readonly Func<T, bool> _shouldReEnqueue;
+
+        /// <summary>
+        /// Initializes the <see cref="T:Orleans.Runtime.TimerWheel"/>.
+        /// </summary>
+        /// <param name="timerPeriod">Period for the internal timer to check the queue.</param>
+        /// <param name="shouldReEnqueue"> If provided will be executed after entry on timeout function.</param>
+        /// <param name="shouldDequeue"> If provided will be executed before the check for timeout. 
+        /// The latter will not be executed if shouldDequeue func returned true</param>
+        public TimerWheel(TimeSpan timerPeriod, Func<T, bool> shouldReEnqueue = null, Func<T, bool> shouldDequeue = null)
+        {
+            _shouldDequeue = shouldDequeue;
+            _shouldReEnqueue = shouldReEnqueue;
+            _queueChecker = new Timer(state =>
+            {
+                CheckQueue();
+            }, null, timerPeriod, timerPeriod);
+        }
+
+        public void Register(T element)
+        {
+            try
+            {
+                _queueLock.Get();
+                _entries.Enqueue(element);
+            }
+            finally
+            {
+                _queueLock.Release();
+            }
+        }
+
+        // Crawls through the callbacks and timeouts expired ones
+        private void CheckQueue()
+        {
+            var now = DateTime.UtcNow;
+            while (true)
+            {
+                try
+                {
+                    _queueLock.Get();
+                    for (int i = 0; i < MaxItemsPerLock; i++)
```

Comment:
There is sleep in between - this  logic is being executed in thread pool with period of 70 ms. MaxItemsPerLock was added to allow other thread make progress, but it seems that it's better to remove it.
In order to get this ordered by time with top of the queue is the earliest I've created [two timer wheels ](https://github.com/dotnet/orleans/pull/2159/files#diff-24969f74ffdd4f9fe9cc680c955cc0ceR38) - one for ordinary messages, and one for resendable as their periods [differs](https://github.com/dotnet/orleans/blob/master/src/Orleans/Runtime/CallbackData.cs#L76).